pingcap/tiflash
# optimize remote read by push filter to remote node

### What problem does this PR solve?

Issue Number: close #1850 

Problem Summary:
As the issue described.
### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
1. add a `force_encode_type` flag in DAGRequest: pingcap/tipb#238
2. set `forc_encode_type` flag to true for coprocessor request between TiFlash nodes.
3. push `selection` to remote read, set correct timezone name and timezone offset in DAGRequest
4. skip timezone cast and selection for data that coming from `CoprocessorBlockInputStream`
### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch:

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects
- Breaking backward compatibility
   after this pr, when there is a TiFlash(client) wants to use `CoprocessorBlockInputStream` to read data from other TiFlash(server), then
  * server is new version, client is old version: everything is fine
  * server is old version, client is new version: client push down filter to the server, server will filter the data as expected, however, if 1. server decide to use `TypeDefault` as the returning encode type(which is false in most scenarios), 2. the timezone is not UTC, 3. the return columns contain timestamp column, then the timestamp column returned by old server will not be added the timezone offset. In this case `CoprocessorReader` will throw exception directly.


### Release note 

```release-note
Please add a release note, or a 'None' if it is not needed.
```
